contracts/upgradeable_contracts/omnibridge_nft/components/common/BridgeOperationsStorage.sol

```diff
@@ -8,53 +8,39 @@ import "../../../../upgradeability/EternalStorage.sol";
     }
 
     /**
-     * @dev Tells the bridged token address of a message sent to the AMB bridge.
-     * @return address of a token contract.
+     * @dev Tells the bridged message checksum.
+     * @param _messageId id of the sent AMB message.
+     * @return saved message checksum associated with the given message id.
      */
-    function messageToken(bytes32 _messageId) internal view returns (address) {
-        return addressStorage[keccak256(abi.encodePacked("messageToken", _messageId))];
+    function getMessageChecksum(bytes32 _messageId) internal view returns (bytes32) {
+        return bytes32(uintStorage[keccak256(abi.encodePacked("messageChecksum", _messageId))]);
     }
 
     /**
-     * @dev Stores the value of a message sent to the AMB bridge.
-     * @param _messageId of the message sent to the bridge.
-     * @param _value amount of tokens bridged.
+     * @dev Calculates message checksum, used for verifying correctness of the given parameters when fixing message.
+     * @param _token address of the bridged token contract on this side current side of the bridge.
+     * @param _sender address of the tokens sender.
+     * @param _tokenIds list of ids of sent tokens.
+     * @param _values list of sent token amounts. Should be an empty array for ERC721 tokens.
+     * @return message checksum.
      */
-    function setMessageValue(bytes32 _messageId, uint256 _value) internal {
-        uintStorage[keccak256(abi.encodePacked("messageValue", _messageId))] = _value;
-    }
-
-    /**
-     * @dev Tells the amount of tokens of a message sent to the AMB bridge.
-     * @return value representing amount of tokens.
-     */
-    function messageValue(bytes32 _messageId) internal view returns (uint256) {
-        return uintStorage[keccak256(abi.encodePacked("messageValue", _messageId))];
-    }
-
-    /**
-     * @dev Stores the receiver of a message sent to the AMB bridge.
-     * @param _messageId of the message sent to the bridge.
-     * @param _recipient receiver of the tokens bridged.
-     */
-    function setMessageRecipient(bytes32 _messageId, address _recipient) internal {
-        addressStorage[keccak256(abi.encodePacked("messageRecipient", _messageId))] = _recipient;
-    }
-
-    /**
-     * @dev Tells the receiver of a message sent to the AMB bridge.
-     * @return address of the receiver.
-     */
-    function messageRecipient(bytes32 _messageId) internal view returns (address) {
-        return addressStorage[keccak256(abi.encodePacked("messageRecipient", _messageId))];
+    function _messageChecksum(
+        address _token,
+        address _sender,
+        uint256[] memory _tokenIds,
+        uint256[] memory _values
+    ) internal pure returns (bytes32) {
+        bytes32 hash1 = keccak256(abi.encodePacked(_tokenIds));
+        bytes32 hash2 = keccak256(abi.encodePacked(_values));
+        return keccak256(abi.encodePacked(_token, _sender, hash1, hash2));
```

Comment:
is it possible to avoid making two additional keccak operations to get `hash1` and `hash2`?
```
keccak256(abi.encodePacked(_token, _sender, _tokenIds, _values))
```
or
```
keccak256(abi.encodePacked(_token, _sender, abi.encodePacked(_tokenIds), abi.encodePacked(_values)))
```

Reply:
I am worried that in such case it will be possible to manipulate values in the dynamical-sized arrays. For example, encodings of the following messages will be the same:
1) `_token = A`, `_sender = B`, `_tokenIds = [1, 2, 3, 4]`, `_values = []`
2) `_token = A`, `_sender = B`, `_tokenIds = [1, 2]`, `_values = [3, 4]`

I don't see a direct way of exploring this, but it seems unsafe to me. Alternatively, we can use just `keccak256(abi.encode(_token, _sender, _tokenIds, _values))`, it will return uniquely decodable result.

Reply:
> we can use just keccak256(abi.encode(_token, _sender, _tokenIds, _values)), it will return uniquely decodable result.

this looks good for me